$MirOS: ports/emulators/qemu/snapshot/patches/patch-cpu-all_h,v 1.1 2008/05/07 15:53:09 tg Exp $
$OpenBSD: patch-cpu-all_h,v 1.3 2008/04/28 22:52:38 todd Exp $
--- cpu-all.h.orig	Wed Sep 23 19:01:07 2009
+++ cpu-all.h	Sat Oct 17 16:35:43 2009
@@ -1034,13 +1034,22 @@ static inline int64_t cpu_get_real_ticks
 }
 
 #else
-/* The host CPU doesn't have an easily accessible cycle counter.
-   Just return a monotonically increasing value.  This will be
-   totally wrong, but hopefully better than nothing.  */
+# warning non-optimized CPU
+#include <sys/time.h>
+#include <time.h>
+
 static inline int64_t cpu_get_real_ticks (void)
 {
-    static int64_t ticks = 0;
-    return ticks++;
+    struct timeval tv;
+    static int64_t i = 0;
+    int64_t j;
+
+    gettimeofday(&tv, NULL);
+    do {
+        j = (tv.tv_sec * (uint64_t) 1000000) + tv.tv_usec;
+    } while (i == j);
+    i = j;
+    return j;
 }
 #endif
 
